src/docs/stan-reference/functions.tex

```diff
@@ -3424,6 +3424,127 @@ \chapter{Mixed Operations}\label{mixed-operations.chapter}
 %
 \end{description}
 
+\chapter{Algebraic Equation Solver}
+\label{functions-algebraic-solver.chapter}
+
+Stan provides a built-in algebraic equation solver. Although it 
+looks like other function applications, the algebraic solver is
+special in two ways.
+
+First, the algebraic solver is a higher-order function, i.e. it takes
+another function as one of its arguments. The only other functions
+in Stan which share this feature are the ordinary differential equation
+solvers (see section~\ref{functions-ode-solver.chapter}). Ordinary Stan
+functions do not allow functions as arguments.
+
+Second, some of the arguments of the algebraic solvers are restricted
+to data only expressions. These expressions must not contain variables
+other than those declared in the data or transformed data blocks. Ordinary
+Stan functions place no restriction on the origin of variables in their argument
+expressions.
```

Comment:
The ODE solvers place these restrictions, but those are also not ordinary functions.

Mitzi's finishing the pull request to have a way to mark user-defined functions as having data-only arguments so that these functions can be called safely within user-defined functions with compile-time rather than run-time errors.